docs/standard/native-interop/tutorial-custom-marshaller.md

```diff
@@ -0,0 +1,298 @@
+---
+title: Use custom marshallers in source-generated P/Invokes
+description: Learn how to use the CustomMarshallerAttribute and implement a custom marshaller for use with source generation.
+author: AaronRobinsonMSFT
+ms.date: 08/22/2022
+helpviewer_keywords:
+  - "Interop"
+  - "Marshalling"
+  - "NativeAOT"
+---
+
+# Tutorial: Use custom marshallers in source-generated P/Invokes
+
+In this tutorial, you'll learn how to implement a marshaller and use it for [custom marshalling](custom-marshalling-source-generation.md) in [source-generated P/Invokes](pinvoke-source-generation.md).
+
+You'll implement marshallers for a built-in type, customize marshalling for a specific parameter and a user-defined type, and specify default marshalling for a user-defined type.
+
+All source code used in this tutorial is available in the [dotnet/samples repository](https://github.com/dotnet/samples/tree/main/core/interop/source-generation/custom-marshalling).
+
+## Overview of the `LibraryImport` source generator
+
+The [`System.Runtime.InteropServices.LibraryImportAttribute`][api_libraryimportattribute] type is the user entry point for a source generator introduced in .NET 7. This source generator is designed to generate all marshalling code at compile time instead of at run time. This has historically been done using `DllImport`, but that approach comes with costs that may not always be acceptable&mdash;see [here][pinvoke_source_generation] for more details. The `LibraryImport` source generator can generate all marshalling code and remove the run-time generation requirement intrinsic to `DllImport`.
```

Comment:
```suggestion
The [`System.Runtime.InteropServices.LibraryImportAttribute`][api_libraryimportattribute] type is the user entry point for a source generator introduced in .NET 7. This source generator is designed to generate all marshalling code at compile time instead of at run time. This has historically been done using `DllImport`, but that approach comes with costs that may not always be acceptable&mdash;see [P/Invoke source generation][pinvoke_source_generation] for more details. The `LibraryImport` source generator can generate all marshalling code and remove the run-time generation requirement intrinsic to `DllImport`.
```